### ISSUE TITLE
Fuzzy Crafting and Providing

### ISSUE THREAD

**Issue by matijas1** [RS485/LogisticsPipes#979]
So i wanted to setup automated fuzzy crafting for the Infinity Skyblock pack.I set up fuzzy crafting tables and pipes with fuzzy upgrades and imported the recipes needed to make a Logistics Crafting Table. (Chest, Crafting Table, Sticks, Wooden Gear,, Stone Gear)
I had 1 oak wood plank and a lot of jungle wood planks in a chest.
On first try to craft 4 of them it said missing 76 oak wood planks(OreDict, IgnDmg, IgnNBT)
So i tried moving the jungle planks to another chest and it missed less planks so i returned them back and removed the oak wood plank and the crafting worked.
So basically when it requests fuzzy crafting it looks trough a chest finds the first item that matches the description and stops looking for other matching items, in this case the jungle planks.
Is there any way you could make it look for all of the matching items because that is kinda annoying xD

**Comment by SonOfTheStars** [comment]
I cant confirm this yet...
For me, it has worked all the time.
Ill test this further now

**Comment by matijas1** [comment]
![image](https://cloud.githubusercontent.com/assets/6332092/14611158/2951ca5a-0591-11e6-823e-150c7e002991.png)
![image](https://cloud.githubusercontent.com/assets/6332092/14611189/46dbd30e-0591-11e6-9c94-2836896a244d.png)
![image](https://cloud.githubusercontent.com/assets/6332092/14611205/5adb099c-0591-11e6-9381-89e2d7247d27.png)
![image](https://cloud.githubusercontent.com/assets/6332092/14611228/67c7c406-0591-11e6-94e7-64ef504b3e92.png)

Some screenshots to make it easier

**Comment by hron84** [comment]
A bit more clarification: as I see the problem is not about the item matching but about how recipes are handled. The crafting table checks the network for an item stack that contains the exactly same amount of the ingredient that is missing. It does not try to think out of the box and try to fullfill requirements based on two or more different (but still oredic compatible) stacks. I ran into this same issue on Infinity Evolved (not Skyblock). The <del>solution</del> workaround was tidy up the tree farm and employ only one wood for crafting.

**Comment by SonOfTheStars** [comment]
Honestly though I feel the Crafting system really has to look for EVERYTHING available that matches before issuing the Request. Not just stop at the first vaiable Node in the Provider System it finds.

And YES I can now confirm this Bug! Even though its not really a Bug more like a Design Flaw.

**Comment by SonOfTheStars** [comment]
Quick Question: What build numbr of LP is FTB Infinity on?
Because I just assembled a quick setup of Drawers+Crafting, Crafted a Chest with Oredicted Wood Planks in my private Instance, had 3 Oak and 5 Birch Planks available and worked JUST FINE.

Sitting on latest build.

@matijas1, IS BOTH the Crafting Table AND the Pipe, wich i hope has a Fuzzy Upgrade, set to have OreDict sync Enabled on each ghost slot of the recipe that uses wood?

**Comment by hron84** [comment]
FTB Infinity Expert Skyblock 1.0.4 is on build 112 according to bug 980. Skyblock 1.1.0 is uses 114.

**Comment by SonOfTheStars** [comment]
Its most likely this isnt a bug but rather a flaw in the users Configuration. I just confirmed it working with build 112 as well!

**Comment by bziemons** [comment]
@davboecki It is not that easy on how that works now.
Having a crafting pipe on something with only one slot will horribly fail with the ore dictionary and the request can never complete.

**Comment by SonOfTheStars** [comment]
@theZorro266 Thats not the case for logistics crafting tables though. What the OP talks about is crafting tables. 

Ive tested the setup with Chests, Barrels, TE Caches, Drawers, DSUs and Ironchests as backend structure. 
Cannot confirm any issues with Standart crafting.

Machines and other single slot inventories to be tested tomorrow.

**Comment by TST-R** [comment]
I tested this on build 114 (crafting planks from logs using a fuzzy crafting table & fuzzy pipe) and the fuzzy crafting is not working for me either. I tried to just connect the fuzzy pipe to a chest, with a regular oak wood input & output configured, and I could not get it to put in other wood types (I tried all kinds of fuzzy modes).

**Comment by DDzwiedziu** [comment]
I'll confirm this on build 114 with two cases: planks and TE gears (iron ingot + 4 desired metal ingots). 
In the second case the system tried to perform the crafting, but for i.e. Forestry copper (set as the default in the recipe, but OreDict'ed) it tried to use Forestry bronze, or, when I removed bronze from the system, it used gold nuggets [SIC!].

**Comment by SonOfTheStars** [comment]
@TST-R @DDzwiedziu I cant confirm this, Every time i do Fuzzy Crafting, it works fine. With every kind of Backend...Be it TE-Caches, Ironchest, Vanilla Chest, AE, Storage Drawers, Jabba, DSU...

**Are you 100% SURE that your Crafting Pipe has a Fuzzy Upgrade, the Table is Fuzzy as well and both are 100% configured the same way?**

A good way to ensure it working is: Place a Fuzzy Table -> Place a Crafting Pipe and Upgrade it to Fuzzy with a Fuzzy Upgrade -> Put the Recipe into the table and select **OreDict ONLY** on the wanted Items -> Hit the "Import" Button on the Crafting Pipes GUI

**Comment by TST-R** [comment]
Yes, I am pretty sure I have tested it exactly like that. Is it possible that there is some dependency/compatibility problem involved?

**Comment by DDzwiedziu** [comment]
I can confirm from memory for now. My world is stuck in a "crashloop".

A new thing is that grass blocks (new in the system) have started to be interpreted as OreDict'ed spruce wood.

Edit: my world is up and the problem persists.
Screenshots
The iron nuggets in the recepie are from Thaumcraft, avaiable (trough Storage Drawers' Compacting ones) are Tinkers' Construct's.
Last screenshot shows what happens if I change the nuggets to TC's.
Edit 2: "hardcoding" the nuggets to TC ones (or any other even) in a regular LP crafting table returns the message that those are missing.
Edit 3 (maybe the last one): Setting the same recepie trough NEI, with the TC nuggets, in a Logistic Request Table, and clicking "~" yields all the components correctly. 

![2016-05-03_21 20 42](https://cloud.githubusercontent.com/assets/3499420/14995485/5a8c2382-1175-11e6-85c5-8d33aa9e6899.png)
![2016-05-03_21 20 44](https://cloud.githubusercontent.com/assets/3499420/14995487/5ab55cc0-1175-11e6-9fac-09f9906d92b0.png)
![2016-05-03_21 20 48](https://cloud.githubusercontent.com/assets/3499420/14995488/5ab65fc6-1175-11e6-9c79-df1d8a08b958.png)
![2016-05-03_21 20 54](https://cloud.githubusercontent.com/assets/3499420/14995489/5ab6dfc8-1175-11e6-9857-27b45f0c52e4.png)
![2016-05-03_21 26 01](https://cloud.githubusercontent.com/assets/3499420/14995586/d8ede094-1175-11e6-9b6d-812b479f94c6.png)

**Comment by ArmaWizard** [comment]
I can confirm while oredictionarying normal wood into wood planks. I get dirt delivered when it is available.
This is still on "Needs-validation" not sure if it is still relevant.

This happens with a crafting pipe + fuzzy upgrade + fuzzy crafting table.
When using a chassis pipe with a crafting module + fuzzy upgrade + fuzzy crafting table this does not happen. But still it does not craft the wood planks; unless I am wrong and I dont fully understand ore dictionarying.

**Comment by bziemons** [comment]
It probably still is relevant

**Comment by ArmaWizard** [comment]
Alright, I don't know if this will be fixed in the milestone; or if this can be hotfixed. Because this severally breaks the auto-crafting if you dont use all the basic minecraft items like oak wood planks. 

Note that I extremely like this work and you guys are doing this in your free time, i'm just happy this mod is still going. Played it as one of the first ones together with buildcraft!

**Comment by bziemons** [comment]
> Because this severally breaks the auto-crafting if you dont use all the basic minecraft items like oak wood planks.

I don't plan to touch this part of the code in the near future. I'll review and accept any pull requests - as always.
